types/a-star/index.d.ts

```diff
@@ -0,0 +1,42 @@
+export = aStar;
+
+declare function aStar<T>(options: aStar.Options<T>): aStar.Result<T>;
```

Comment:
the obvious problem here is this accepts anything, say number type, while the node is  a shape, with at least `x`, `y` and `toString` available (as per inline docs). Using some constraint on type can be beneficial for users IMO

Reply:
That's not really the case, nodes don't have to have `x` or `y`, or even a `toString()` function available (that's why the `astar()` function accepts a custom toString implementation via the `hash` argument for node types that don't provide one). Plain number types work too, and even `null` and `undefined` technically work even though they obviously don't make much sense.

For example:
```js
astar({
    start: 10,
    isEnd: (key) => key === 0,
    neighbor: (key) => [key - 1, key + 1],
    distance: (key1, key2) => Math.abs(key1 - key2),
    heuristic: (key) => key,
});
// { status: 'success', cost: 10, path: [10, 9, 8, 7, 6, 5, 4, 3, 2, 1, 0] }
```

```js
astar({
    start: 10,
    isEnd: (key) => key === null,
    neighbor: (key) => key === 0 ? [null] : [key - 1, key + 1],
    distance: (key1, key2) => 1,
    heuristic: (key) => key,
    hash: (key) => key === null ? 'null' : key.toString(),
});
// { status: 'success', cost: 11, path: [10, 9, 8, 7, 6, 5, 4, 3, 2, 1, 0, null] }
```

Reply:
oh, yes, you're correct. Also just noticed the last line in readme about unrestricted data types. 
But that does not mean, it would serve the best TS purpose (helping to code not failing code). For example, if those types like `null`, `undefined` are used, users have to provide hash function, otherwise the builtin one will throw error at runtime. Use the snip, just skip optional hash:
```js
astar({
  start: 10,
  isEnd: (key) => key === null,
  neighbor: (key) => key === 0 ? [null] : [key - 1, key + 1],
  distance: (key1, key2) => 1,
  heuristic: (key) => key,
});
```
```js
TypeError: Cannot read properties of null (reading 'toString')
```
that was my point when commenting in first place.